cdap-security/src/main/java/io/cdap/cdap/security/authorization/InternalAccessEnforcer.java

```diff
@@ -0,0 +1,106 @@
+
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.Set;
+
+public class InternalAccessEnforcer extends AbstractAccessEnforcer {
+  private static final Logger LOG = LoggerFactory.getLogger(InternalAccessEnforcer.class);
+
+  private final TokenManager tokenManager;
+  private final Codec<AccessToken> accessTokenCodec;
+
+  @Inject
+  InternalAccessEnforcer(CConfiguration cConf, TokenManager tokenManager, Codec<AccessToken> accessTokenCodec) {
+    super(cConf);
+    this.tokenManager = tokenManager;
+    this.accessTokenCodec = accessTokenCodec;
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal,
+                      Set<? extends Permission> permissions) throws AccessException {
+    validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+  }
+
+  @Override
+  public void enforceOnParent(EntityType entityType, EntityId parentId, Principal principal,
+                              Permission permission) throws AccessException {
+    validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+  }
+
+  @Override
+  public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds,
+                                           Principal principal) throws AccessException {
+    try {
+      validateAccessTokenAndIdentity(principal.getName(), principal.getFullCredential());
+      return entityIds;
+    } catch (AccessException e) {
+      return Collections.emptySet();
+    }
+  }
+
+  private void validateAccessTokenAndIdentity(String principalName, Credential credential) throws AccessException {
+    if (credential == null) {
+      throw new IllegalStateException("Attempted to internally enforce access on null credential");
+    }
+    if (!credential.getType().equals(Credential.CredentialType.INTERNAL)) {
+      throw new IllegalStateException("Attempted to internally enforce access on non-internal credential type");
+    }
+    AccessToken accessToken;
+    try {
+      accessToken = accessTokenCodec.decode(Base64.getDecoder().decode(credential.getValue()));
+    } catch (IOException e) {
+      throw new AccessException("Failed to deserialize access token", e);
+    }
+    try {
+      tokenManager.validateSecret(accessToken);
+    } catch (InvalidTokenException e) {
+      throw new AccessException("Failed to validate access token", e);
+    }
+    UserIdentity userIdentity = accessToken.getIdentifier();
+    if (!userIdentity.getUsername().equals(principalName)) {
+      LOG.debug(String.format("Internal access token username differs from principal name; got token " +
+                                                "name '%s', expected principal name '%s'",
+                                              userIdentity.getUsername(), principalName));
+    }
+    if (userIdentity.getIdentifierType() == null || !userIdentity.getIdentifierType()
+      .equals(UserIdentity.IdentifierType.INTERNAL)) {
```

Comment:
same as above

Reply:
Since this is a cherry-pick, I've gone ahead and opened a new PR #13533 to address these comments.